### PR TITLE
Use libclang to parse code

### PR DESCRIPTION
**Pull request by Javier-varez**
Added as an option and kept ctags as a default, which might be
needed for golang code

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -16,7 +16,10 @@ jobs:
       with:
         go-version: '1.17' # The Go version to download (if necessary) and use.
     - run: sudo apt install -y universal-ctags pandoc
-    - run: go test ./...
+    - run: mkdir .tools && wget https://github.com/daedaleanai/llvm-project/releases/download/ddln-llvm14-rc/llvm-14.0.0.tar.gz && tar -xf llvm-14.0.0.tar.gz -C .tools
```
Comment (finomen) [inline .github/workflows/main.yml:19]: Github workflows has weird free quota and pricing policy if free quota exceeded. Let's check that downloading gigabytes of data will not cost a fortune

File: config/config.go [hunk 2]
```diff
@@ -40,8 +40,10 @@ type jsonFileQuery struct {
 }
 
 type jsonImplementation struct {
-	Code  jsonFileQuery `json:"code"`
-	Tests jsonFileQuery `json:"tests"`
+	Code                jsonFileQuery `json:"code"`
+	Tests               jsonFileQuery `json:"tests"`
+	CompilationDatabase string        `json:"compilationDatabase"`
+	ClangArguments      string        `json:"clangArguments"`
```
Comment (finomen) [inline config/config.go:46]: Why this is not array?

File: testdata/libclangtest/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.16)
+
+project(myproj)
+
+add_executable(myexec code/a.cc)
+target_include_directories(myexec PRIVATE code/include)
+target_compile_options(myexec PRIVATE -Os -std=c++17)
```
Comment (finomen) [inline testdata/libclangtest/CMakeLists.txt:7]: set_target_properties(myexec PROPERTIES
    CXX_STANDARD 20
    CXX_STANDARD_REQUIRED ON
)

File: testdata/libclangtest/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+add_executable(myexec_tests test/a/a_test.cc)
+target_include_directories(myexec_tests PRIVATE code/include)
+target_compile_options(myexec_tests PRIVATE -Os -std=c++17)
```
Comment (finomen) [inline testdata/libclangtest/CMakeLists.txt:11]: set_target_properties(myexec_tests PROPERTIES
    CXX_STANDARD 20
    CXX_STANDARD_REQUIRED ON
)

**Review by finomen — CHANGES_REQUESTED** [review]

File: config/config.go (hunk 2 above)
Comment (Javier-varez) [reply config/config.go:46]: Just to make it easier for the user to write it as a regular string really. I can change it if you would prefer it as an array.

File: .github/workflows/main.yml (hunk 1 above)
Comment (Javier-varez) [reply .github/workflows/main.yml:19]: What I read is that the pricing policy only applies for private repositories, which this is not, so I think this should be fine. 

> GitHub Actions usage is free for both public repositories and self-hosted runners. For private repositories, each GitHub account receives a certain amount of free minutes and storage, depending on the product used with the account. Any usage beyond the included amounts is controlled by spending limits.

Also it seems that by default there is an option to set limits to the expenses (and for monthly billed customers the default is zero):
> If you are a monthly-billed customer, your account will have a default spending limit of 0 US dollars (USD), which prevents additional usage of minutes or storage for private repositories beyond the amounts included with your account. If you pay your account by invoice, your account will have an unlimited default spending limit. For more information, see "[About spending limits](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#about-spending-limits)."

Lastly, they seem to charge based on minutes of execution and storage. I didn't read anything about network bandwidth or usage.

https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions

So I think this should be fine

Comment (Javier-varez) [reply .github/workflows/main.yml:19]: Sent an internal email to check the policy, in the meantime, I think we can go ahead with this.

**Review by finomen — APPROVED** [review]

File: certdocs/TRAQ-138-SDD.md [hunk 5]
```diff
@@ -997,6 +997,23 @@ Reqtraq SHALL provide a command line option to start a local web sever to genera
 - Verification: Test
 - Safety Impact: None
 
+### clang.go
+
+Reqtraq can use libclang to parse the ast of any linked code and obtain code references. This option
+is preferred for a fine-grained code parsing, while ctags is preferred for compatibility with other 
+languages like Go.
+
+#### REQ-TRAQ-SWL-61 Support for libclang backend to parse code
```
Comment (dawzee) [inline certdocs/TRAQ-138-SDD.md:1006]: I feel this deserves more than one LLR. There's some derived functionality in the code which should be specified in low-level requirements (along with rationales), e.g. the skipping of certain elements, the use of command database

File: .github/workflows/main.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Javier-varez) [reply .github/workflows/main.yml:19]: Generally speaking, third-party users would use llvm-14 from upstream, no need for daedaleans. However, for our usecase it makes sense to use this one, because it differs a bit from upstream llvm and we have some fixes, so in order to properly parse our code, we should use this version.

Regarding the setup_workspace,  I'm not quite sure how that works, but I think the idea was to integrate reqtraq with dbt as a dependency, where we could specify the pinned llvm versions we have and where they are (given llvm is also a dbt dependency).

File: testdata/libclangtest/reqtraq_config.json [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+                "code": {
+                    "paths": ["code"],
+                    "matchingPattern": ".*\\.(cc|hh)$",
+                    "ignoredPatterns": [".*_ignored\\.(cc|hh)$"]
+                },
+                "tests": {
+                    "paths": ["test"],
+                    "matchingPattern": ".*_test\\.(cc|hh)$"
+                },
+                "clangArguments": [
```
[the remark this answers is not in the document]
Comment (Javier-varez) [reply testdata/libclangtest/reqtraq_config.json:49]: Yes, I can add that, and maybe track the compilation database instead of generating it with  cmake each time

File: testdata/libclangtest/code/include/a.hh [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+     */
+    T &operator[](std::size_t index) { return mData[index]; }
+
+   private:
+    T mData[N];
+
+    /*
+     * This one should not appear, since it is private
+     */
+    void MorePrivateStuff();
```
[the remark this answers is not in the document]
Comment (Javier-varez) [reply testdata/libclangtest/code/include/a.hh:53]: As far as I understood (and this is also how most of stdlib looks like now), private code that is not callable by the user (e.g. private), and is considered an implementation detail, not provided functionality. If this is not the case, I can certainly relax this rule and catch all private methods. And of course, add all the missing reqs to stdlib and the rest of the code.

File: testdata/libclangtest/code/include/a.hh [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    void ButCanHavePrivateFunctions();
+};
+
+/**
+ * \llr REQ-TEST-SWL-2
+ */
+void JustAFreeFunction();
+
+namespace {
+void functionInAnAnonymousNamespace();
```
[the remark this answers is not in the document]
Comment (Javier-varez) [reply testdata/libclangtest/code/include/a.hh:78]: Same as above, detail and anonymous namespaces are used for implementation detail that is not directly called by  the user.

File: config/config_test.go [hunk 9]
```diff
@@ -61,8 +62,9 @@ func TestConfig_ParseConfig(t *testing.T) {
 				},
 			},
 			Implementation: Implementation{
-				CodeFiles: []string{},
-				TestFiles: []string{},
+				CodeFiles:           []string{},
+				TestFiles:           []string{},
+				CompilationDatabase: "",
```
[the remark this answers is not in the document]
Comment (Javier-varez) [reply config/config_test.go:67]: No I just forgot, and it works because it's empty anyway

File: clang.go [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+		case clang.Cursor_ClassDecl, clang.Cursor_EnumDecl, clang.Cursor_StructDecl, clang.Cursor_ClassTemplate, clang.Cursor_ClassTemplatePartialSpecialization:
+			if !IsPublic(cursor) {
+				return clang.ChildVisit_Continue
+			}
+
+			// Only recurse if there is a chance that we can find something of interest.
+			// In practice if the element is not public, there is no point in recursing
+			return clang.ChildVisit_Recurse
+
+		case clang.Cursor_Namespace:
```
[the remark this answers is not in the document]
Comment (Javier-varez) [reply clang.go:125]: It is not the same. `IsInAnonymousOrDetailNamespaceOrClass` is checking if a node in the AST is part of an anonymous namespace or detail namespace. This checks if the AST node that represents the current cursor is in itself a declaration of an anonymous or detail namespace and then doesn't recurse into its inner nodes. In reality with this we should not need `IsInAnonymousOrDetailNamespaceOrClass`, but for some reason, I was still getting functions in detail namespaces (I think due to template instantiation, which happens out of the original namespace declaration)

File: clang.go [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+	code := make(map[CodeFile][]*Code)
+
+	index := clang.NewIndex(1, 0)
+	defer index.Dispose()
+
+	var compDb clang.CompilationDatabase
+	if compilationDatabase != "" {
+		pathInRepo, err := repos.PathInRepo(repoName, compilationDatabase)
+		if err != nil {
+			fmt.Printf("Error opening compilation database in path `%s`: `%v`", compilationDatabase, err)
```
[the remark this answers is not in the document]
Comment (Javier-varez) [reply clang.go:236]: I don't want to make this a hard error (although I have been going back and forth on this), because it just means that it will fall back to the clangArguments and could still potentially parse the code. I can rephrase this as a warning though, that would make it clearer.

File: .github/workflows/main.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Javier-varez) [reply .github/workflows/main.yml]: I added a build tag to conditionally add support for libclang. This way, whoever doesn't need libclang, does not need to install llvm (e.g.: prod repos). I think that  resolves this question.

File: testdata/libclangtest/code/include/a.hh (hunk 7 above)
[the remark this answers is not in the document]
Comment (dawzee) [reply testdata/libclangtest/code/include/a.hh:53]: as discussed, the combination of our design standards (LLRs are defined at module level) and coding standards (private methods do not need to be documented) leads to this. I will raise a ticket to make this explicit in the coding standards next time they are updated.

File: testdata/libclangtest/code/include/a.hh (hunk 8 above)
[the remark this answers is not in the document]
Comment (dawzee) [reply testdata/libclangtest/code/include/a.hh:78]: as above, will be made explicit in coding standards at next update

**Review by dawzee — APPROVED** [review]

File: .github/workflows/main.yml [hunk 12]
```diff
@@ -40,4 +55,7 @@ jobs:
       with:
         go-version: '1.17' # The Go version to download (if necessary) and use.
     - run: sudo apt install -y universal-ctags pandoc
+    - run: mkdir .tools && wget https://github.com/daedaleanai/llvm-project/releases/download/ddln-llvm14-rc/llvm-14.0.0.tar.gz && tar -xf llvm-14.0.0.tar.gz -C .tools
```
[the remark this answers is not in the document]
Comment (Javier-varez) [reply .github/workflows/main.yml:58]: ah, good point, will remove this.
